File: command/src/main/java/com/nearsoft/workshop/command/CeilingFan.java
package com.nearsoft.workshop.command;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class CeilingFan {

    // @Slf4j annotation in Lombok or Groovy
    private final Logger log = LoggerFactory.getLogger(this.getClass());

    public enum CeilingFanSpeed {OFF, LOW, MEDIUM, HIGH}

    private CeilingFanSpeed speed;

    public CeilingFan() {
        this.speed = CeilingFanSpeed.OFF;
        log.debug("Created CeilingFan");
    }

    public void high() {
        this.speed = CeilingFanSpeed.HIGH;
        log.debug("CeilingFan setted to HIGH");
    }

    public void medium() {
        this.speed = CeilingFanSpeed.MEDIUM;
        log.debug("CeilingFan setted to MEDIUM");
    }

    public void low() {
        this.speed = CeilingFanSpeed.LOW;
        log.debug("CeilingFan setted to LOW");
    }

    public void off() {
        this.speed = CeilingFanSpeed.OFF;
        log.debug("CeilingFan setted to OFF");
    }
}

File: command/src/main/scala/com/nearsoft/workshop/command/scala/CeilingFan.scala
package com.nearsoft.workshop.command.scala

import org.slf4j.LoggerFactory

class CeilingFan {
  val log = LoggerFactory.getLogger(classOf[CeilingFan])

  object CeilingFanSpeed extends Enumeration {
    type CeilingFanSpeed = Value
    val OFF, LOW, MEDIUM, HIGH = Value
  }

  import CeilingFanSpeed._

  var speed = OFF
  log.debug("Created CeilingFan")

  def high = {
    speed = HIGH
    log.debug("CeilingFan setted to HIGH")
  }

  def medium = {
    speed = MEDIUM
    log.debug("CeilingFan setted to MEDIUM")
  }

  def low = {
    speed = LOW
    log.debug("CeilingFan setted to LOW")
  }

  def off = {
    speed = OFF
    log.debug("CeilingFan setted to OFF")
  }


}
